### PR TITLE
Update MTableHeader to support showDetailPanelIcon

### PR DESCRIPTION
**Pull request by peacechen**
## Related Issue

This is the header portion of #87 

## Description

The previous PR removed the detail panel icon in the body row, but the header continues to render a column for it. That causes the header row to offset incorrectly. This PR conditionally removes the header column corresponding to the detail panel icon.

## Related PRs

#136 

## Impacted Areas in Application

MTableHeader

### CODE REVIEW

**Comment by peacechen** [comment]
@oze4 @Domino987 
It would be great if this were included in the next release. We're hoping to use the new `showDetailPanelIcon` and `renderSummaryRow` features soon.

**Comment by peacechen** [comment]
Update: The summary row was not accounting for the action icons. The column offset was thus incorrect when those are present. I pushed a commit to fix that.

**Comment by peacechen** [comment]
Another corner case is in edit mode. The columns are rendered different in edit mode and don't align with normal mode that the summary row expects. The simplest way to address that for now is to empty the render row when editing. To achieve that, I added a callback prop `onBulkEditOpen` to let the parent component know when the table is in edit mode.

There was also another bug in `MTableEditRow` that caused a warning. It was passing the `bulkEditChangedRows` prop to TableRow and React complained that's not supported.

**Comment by Domino987** [comment]
Thank you for your work. I will take a look at it tomorrow morning.

File: src/components/MTableEditRow/m-table-edit-row.js [hunk 1]
```diff
@@ -171,7 +171,7 @@ export default class MTableEditRow extends React.Component {
 
   renderActions() {
     if (this.props.mode === 'bulk') {
-      return <TableCell padding="none" key="key-actions-column" />;
```
Comment (matthewoestreich) [inline src/components/MTableEditRow/m-table-edit-row.js:174]: Why remove this? Do we not want to return an empty table cell here? I could be wrong - just curious.

@peacechen @Domino987

File: src/components/MTableSummaryRow/index.js [hunk 2]
```diff
@@ -2,19 +2,75 @@ import * as React from 'react';
...
+    if (rowProps.options.detailPanelColumnAlignment === 'right') {
+      placeholderRightColumns.push(renderPlaceholderColumn(placeholderKey++));
+    } else {
+      placeholderLeftColumns.push(renderPlaceholderColumn(placeholderKey++));
+    }
+  }
+  if (rowProps.isTreeData) {
+    placeholderLeftColumns.push(renderPlaceholderColumn(placeholderKey++));
+  }
+
```
Comment (matthewoestreich) [inline src/components/MTableSummaryRow/index.js:70]: I think I see what you're doing here. If I understand this correctly, this solves the issue where sometimes the columns are not aligned correctly with it's header? Should we put this in a different PR, if so?

**Review by matthewoestreich — COMMENTED** [review]
Overall I think it looks good. My only concern is we may want to separate into separate PRs? Not that big of a deal, but that's my only "criticism".

Thank you @peacechen for your contribution/time! Greatly appreciated.

**Comment by Domino987** [comment]
Can we throw this in a sandbox so we can also verify the code? @peacechen I am not sure if all use cases are covered with this.

File: src/components/MTableEditRow/m-table-edit-row.js (hunk 1 above)
Comment (peacechen) [reply src/components/MTableEditRow/m-table-edit-row.js:174]: TableCell takes up space even if its contents are empty. For example when the details icon isn't rendered, the icon is gone, but the TableCell consumes a column's worth of space and incorrectly offsets the header.

File: src/components/MTableSummaryRow/index.js (hunk 2 above)
Comment (peacechen) [reply src/components/MTableSummaryRow/index.js:70]: Yes that's what it fixes.

I would've preferred to separate these into their own PR's, but Github makes that difficult. I would've needed to create separate branches for each one. We needed all these fixes for our product last week and there was no movement on merging PRs.

**Comment by peacechen** [comment]
> Can we throw this in a sandbox so we can also verify the code? @peacechen I am not sure if all use cases are covered with this.

I tested this during development using the demo app in this repo.  I suppose we could add additional copies of the table with different permutations of props. Do you prefer that or adding props to the existing sample tables?

We're using my fork in production and it's working as expected.

**Review by matthewoestreich — APPROVED** [review]
I will go ahead and approve the PR - @Domino987 what do you think?

File: src/components/MTableEditRow/m-table-edit-row.js (hunk 1 above)
Comment (Domino987) [reply src/components/MTableEditRow/m-table-edit-row.js:174]: Yes this looks good to me.

File: src/components/MTableHeader/index.js [hunk 3]
```diff
@@ -272,7 +272,7 @@ export function MTableHeader(props) {
         headers.push(renderActionsHeader());
       }
     }
-    if (props.hasDetailPanel) {
+    if (props.hasDetailPanel && props.options.showDetailPanelIcon) {
```
Comment (Domino987) [inline src/components/MTableHeader/index.js:275]: What is the default here? We don't want to brick existing code.

Comment (peacechen) [reply src/components/MTableHeader/index.js:275]: Default is true
https://github.com/material-table-core/core/blob/master/src/defaults/props.options.js#L34

**Comment by peacechen** [comment]
I have another feature in the pipeline : controlled editing mode.  That allows the parent to put the table in edit mode instead of requiring the user to press the edit button.  It would be preferable for that to be in a separate PR. When will this one be merged so I can submit the next one?

**Comment by matthewoestreich** [comment]
@peacechen isn't that already possible, or am I missing something?

```jsx
function EditableRow(props) {
  const [isEditing, setIsEditing] = useState(false);

  return (
    <div>
      <p>I am the parent</p>
      <button onClick={(e) => setIsEditing(!isEditing)}>
        {isEditing ? 'Disable' : 'Enable'} Editing
      </button>
      <MaterialTable
        editable={{
          isEditable: () => isEditing,
          onRowUpdate: (newData, oldData) =>
            new Promise((resolve, reject) => setTimeout(() => resolve(), 1000))
        }}
        data={[
          { name: 'jack', id: 1 },
          { name: 'nancy', id: 2 }
        ]}
        columns={[
          { field: 'name', title: 'Name' },
          { field: 'id', title: 'Identifier' }
        ]}
      />
    </div>
  );
}
```

**Comment by matthewoestreich** [comment]
@peacechen 

> Another corner case is in edit mode. ...

I think this is why putting the issues in a sandbox makes the most sense. I think I see what you are accomplishing here, but understanding (and seeing) the issue/issues helps us all be on the same page. I think that may help with merge times, etc..

**Comment by Domino987** [comment]
I will merge this. This looks good. Thank you for your work. its appreciated.

**Comment by matthewoestreich** [comment]
Nice! As far as controlled edit mode - I moved that [here](https://github.com/material-table-core/core/discussions/163)

**Comment by peacechen** [comment]
> @peacechen isn't that already possible, or am I missing something?

The goal is to allow the parent to programmatically enter edit mode. Currently the user must click on the edit icon (e.g. the pencil icon). The code that you provided adds the edit icon to each row, but the user must click on the icon. The same is true for bulk edit mode.

File: src/components/MTableEditRow/m-table-edit-row.js (hunk 1 above)
Comment (Llolow) [reply src/components/MTableEditRow/m-table-edit-row.js:174]: I understand when we have icons it's incorrectly offsets the header, but with this modification it's render the bulkEdit verry weird :
Alt Text 
when TableCell is removed the th of Action isn't removed,
 idk if i forgot a thing with you'r new option.
Thx

File: src/components/MTableSummaryRow/index.js (hunk 2 above)
Comment (matthewoestreich) [reply src/components/MTableSummaryRow/index.js:70]: @Llolow I thought this would fix that issue where the body doesn't line up with the headers when editing, but it doesn't look like I was right. That's def something that needs to be fixed though.

File: src/components/MTableEditRow/m-table-edit-row.js (hunk 1 above)
Comment (peacechen) [reply src/components/MTableEditRow/m-table-edit-row.js:174]: Looks like there's a bug for the other actions such as Delete not being accounted for in edit mode. I'll investigate and submit a PR. In the meantime, please paste your comment and screenshot into a new ticket so that it can be referenced by the PR.
